nexus_workcell_orchestrator/src/job_manager.cpp

```diff
@@ -0,0 +1,257 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "job_manager.hpp"
+
+namespace nexus::workcell_orchestrator {
+
+using TaskState = nexus_orchestrator_msgs::msg::TaskState;
+using WorkcellRequest = endpoints::WorkcellRequestAction::ActionType;
+
+/**
+ * Tick the behavior tree of a job and update the task status according to the result.
+ */
+void JobManager::_tick_bt(Job& job)
+{
+  this->_ctx_mgr->set_active_context(job.ctx);
+  const auto bt_status = job.bt->tickRoot();
+  ++job.tick_count;
+  this->_ctx_mgr->set_active_context(nullptr);
+
+  // update task status according to bt status
+  switch (bt_status)
+  {
+    case BT::NodeStatus::RUNNING: {
+      job.task_state.status = TaskState::STATUS_RUNNING;
+      break;
+    }
+    case BT::NodeStatus::FAILURE: {
+      RCLCPP_INFO(
+        job.ctx->node->get_logger(), "Task [%s] failed",
+        job.ctx->task.id.c_str());
+      // Publish feedback.
+      job.task_state.status = TaskState::STATUS_FAILED;
+      auto fb = std::make_shared<WorkcellRequest::Feedback>();
+      fb->state = job.task_state;
+      job.goal_handle->publish_feedback(std::move(fb));
+
+      // Abort the action request.
+      auto result =
+        std::make_shared<endpoints::WorkcellRequestAction::ActionType::Result>();
+      result->success = false;
+      result->message = "Task failed";
+      job.goal_handle->abort(result);
+      break;
+    }
+    case BT::NodeStatus::SUCCESS: {
+      // Publish feedback.
+      job.task_state.status = TaskState::STATUS_FINISHED;
+      auto fb = std::make_shared<WorkcellRequest::Feedback>();
+      fb->state = job.task_state;
+      job.goal_handle->publish_feedback(std::move(fb));
+
+      auto result =
+        std::make_shared<endpoints::WorkcellRequestAction::ActionType::Result>();
+      result->success = true;
+      result->result = YAML::Dump(job.ctx->task.previous_results);
+      auto report = job.bt_logging->generate_report();
+      result->message = common::ReportConverter::to_string(report);
+      RCLCPP_INFO(job.ctx->node->get_logger(), "%s", result->message.c_str());
+      job.goal_handle->succeed(result);
+      break;
+    }
+    default: {
+      std::ostringstream oss;
+      oss << "Error ticking task [" << job.ctx->task.id <<
+        "]: Behavior tree returned invalid or unknown status [" <<
+        bt_status << "]";
+      RCLCPP_ERROR_STREAM(job.ctx->node->get_logger(), oss.str());
+      throw std::runtime_error(oss.str());
```

Comment:
The reason is because `WorkcellOrchestrator` may need to know when an error happens (can't rmb if we actually need it currently but having that option is good). May be a good idea to halt + throw exception, I think either is ok, having workcell orchestrator handle the error does make it easier to handle edge cases.